src/KeyRegistrar.sol

```diff
@@ -0,0 +1,472 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.27;
+
+import {Initializable} from "@openzeppelin-upgrades/contracts/proxy/utils/Initializable.sol";
+import {OwnableUpgradeable} from "@openzeppelin-upgrades/contracts/access/OwnableUpgradeable.sol";
+import {BN254} from "./libraries/BN254.sol";
+
+/**
+ * @title KeyRegistrar
+ * @notice A core singleton contract that manages operator keys for different AVSs
+ * @dev Provides registration, deregistration, and rotation of keys with support for aggregate keys
+ */
+contract KeyRegistrar is Initializable, OwnableUpgradeable {
+    using BN254 for BN254.G1Point;
+
+    /// @dev Returns the hash of the zero pubkey for BN254 aka BN254.G1Point(0,0)
+    bytes32 internal constant ZERO_PK_HASH = hex"ad3228b676f7d3cd4284a5443f17f1962b36e491b30a40b2405849e597ba5fb5";
+
+    /// @dev Enum defining supported curve types
+    enum CurveType {
+        ECDSA,
+        BN254
+        // BLS12_381 // Future support
+    }
+
+    /// @dev Structure to store key information across different curves
+    struct KeyInfo {
+        bool isRegistered;
+        uint256 lastRotationBlock;
+    }
+
+    /// @dev Maps (AVS, operator, curve type) to key info
+    mapping(address => mapping(address => mapping(CurveType => KeyInfo))) private avsOperatorKeyInfo;
+
+    /// @dev Maps (AVS, operator) to their ECDSA public keys
+    mapping(address => mapping(address => bytes)) private avsOperatorToECDSAKey;
+
+    /// @dev Maps (AVS, operator) to their BN254 G1 points
+    mapping(address => mapping(address => BN254.G1Point)) private avsOperatorToBN254Key;
+
+    /// @dev Maps (AVS, operator) to their BN254 G2 points (for BLS verification)
+    mapping(address => mapping(address => BN254.G2Point)) private avsOperatorToBN254KeyG2;
+
+    /// @dev Maps (AVS, operator, curve type) to key hash for quick lookup
+    mapping(address => mapping(address => mapping(CurveType => bytes32))) private avsOperatorToKeyHash;
+
+    /// @dev Maps (AVS, curve type, key hash) to operator address for reverse lookup
+    mapping(address => mapping(CurveType => mapping(bytes32 => address))) private avsKeyHashToOperator;
+
+    /// @dev Maps (AVS, operatorSetId) to their aggregate BN254 G1 point
+    mapping(address => mapping(uint32 => BN254.G1Point)) private avsOperatorSetToAggregateBN254Key;
```

Comment:
why store the aggregate in the contract instead of calculating it in the calculator?

Reply:
this would introduce an "updateOperators" like pattern for maintaining the apk... imo better to do it in here , but keep it in an `extraBytes` field